server/utils/licenceStatus.js

```diff
@@ -0,0 +1,324 @@
+const {taskStates} = require('../models/taskStates');
+const {licenceStages} = require('../models/licenceStages');
+
+const {matcher, isPresent, isNotPresent, equalTo} = require('./jsonUtils');
+
+module.exports = {getLicenceStatus};
+
+function getLicenceStatus(licenceRecord) {
+
+    if (!licenceRecord || !licenceRecord.licence || !licenceRecord.status) {
+        return {
+            stage: licenceStages.UNSTARTED,
+            decisions: {},
+            tasks: {
+                exclusion: taskStates.UNSTARTED,
+                crdTime: taskStates.UNSTARTED,
+                suitability: taskStates.UNSTARTED,
+                eligibility: taskStates.UNSTARTED,
+                optOut: taskStates.UNSTARTED,
+                bassReferral: taskStates.UNSTARTED,
+                curfewAddress: taskStates.UNSTARTED,
+                riskManagement: taskStates.UNSTARTED,
+                curfewAddressReview: taskStates.UNSTARTED,
+                curfewHours: taskStates.UNSTARTED,
+                reportingInstructions: taskStates.UNSTARTED,
+                licenceConditions: taskStates.UNSTARTED,
+                approval: taskStates.UNSTARTED
+            }
+        };
+    }
+    const stage = licenceRecord.status;
+    const data = matcher(licenceRecord.licence);
+
+    const results = getRequiredState(stage, data);
+    const licenceStatus = results.reduce(combiner, {stage, decisions: {}, tasks: {}});
+
+    console.log(licenceStatus);
+    return licenceStatus;
+}
+
+function getRequiredState(stage, data) {
+
+    switch (stage) {
+        case licenceStages.DECIDED:
+        // fall through
+        case licenceStages.APPROVAL:
+            return [getEligibilityStageState(data), getRoStageState(data), getApprovalStageState(data)];
+        case licenceStages.PROCESSING_CA:
+        // fall through
+        case licenceStages.PROCESSING_RO:
+            return [getEligibilityStageState(data), getRoStageState(data)];
+        case licenceStages.ELIGIBILITY:
+            return [getEligibilityStageState(data)];
+        default:
+            return [];
+    }
```

Comment:
Don't object to this method. If it were me, I'd do something like: 
```
function getRequiredState(stage, data) {

    const config = {
        [licenceStages.DECIDED]: [getEligibilityStageState, getRoStageState, getApprovalStageState],
        [licenceStages.APPROVAL]: [getEligibilityStageState, getRoStageState, getApprovalStageState],
        [licenceStages.PROCESSING_CA]: [getEligibilityStageState, getRoStageState],
        [licenceStages.PROCESSING_RO]: [getEligibilityStageState, getRoStageState],
        [licenceStages.ELIGIBILITY]: [getEligibilityStageState],
    };

    return config[stage].map(getStateMethod => getStateMethod(data));
}

```

Reply:
Yeah that's nicer